src/hotspot/share/opto/type.hpp

```diff
@@ -570,99 +580,179 @@ class TypeInteger : public Type {
+ *
+ * The last 2 conditions mean that the bits in t3._bits._zeros is not a subset
+ * of those in t2._bits._zeros, the same applies to _bits._ones
+ *
+ * As a result, every TypeInt is canonicalized to its optimal form upon
+ * construction. This makes it easier to reason about them in optimizations.
+ * E.g a TypeInt t with t._lo < 0 will definitely contain negative values. It
+ * also makes it trivial to determine if a TypeInt instance is a subset of
+ * another.
+ *
+ * Properties:
+ *
+ * 1. Since every TypeInt instance is canonicalized, all the bounds must also
+ * be elements of such TypeInt. Or else, we can tighted the bounds by narrowing
+ * it by one, which contradicts the assumption of the TypeInt being canonical.
+ *
+ * 2. Either _lo == jint(_lo) and _hi == jint(_uhi), or all elements of a
+ * TypeInt lie in the intervals [_lo, jint(_uhi)] or [jint(_ulo), _hi]
+ *
+ * Proof: For 2 jint value x, y such that they are both >= 0 or < 0. Then:
+ *
+ * x <= y iff juint(x) <= juint(y)
+ *
+ * Then, we have:
+ *
+ * For a TypeInt t, there are 3 possible cases:
+ *
+ * a. t._lo >= 0. Since 0 <= t._lo <= jint(t._ulo), we have:
+ *
+ * juint(t._lo) <= juint(jint(t._ulo)) == t._ulo <= juint(t._lo)
+ *
+ * Which means that t._lo == jint(t._ulo). Similarly, t._hi == jint(t._uhi).
+ *
+ * b. t._hi < 0. Similarly, t._lo == jint(t._ulo) and t._hi == jint(t._uhi)
+ *
+ * c. t._lo < 0, t._hi >= 0. Then jint(t._ulo) >= 0 and jint(t._uhi) < 0. In
+ * this case, all elements of t belongs to either [t._lo, jint(t._uhi)] or
+ * [jint(t._ulo), t._hi].
+ *
+ * This property is useful for our analysis of TypeInt values. Additionally, it
+ * can be seen that _lo and jint(_uhi) are both < 0 or >= 0, and the same
+ * applies to jint(_ulo) and _hi.
+ */
 class TypeInt : public TypeInteger {
-  TypeInt( jint lo, jint hi, int w );
+  TypeInt(const TypeIntPrototype<jint, juint>& t, int w, bool dual);
+  static const Type* try_make(const TypeIntPrototype<jint, juint>& t, int w, bool dual);
 protected:
-  virtual const Type *filter_helper(const Type *kills, bool include_speculative) const;
+  virtual const Type* filter_helper(const Type* kills, bool include_speculative) const;
 
 public:
   typedef jint NativeType;
-  virtual bool eq( const Type *t ) const;
+  virtual bool eq(const Type* t) const;
   virtual uint hash() const;             // Type specific hashing
   virtual bool singleton(void) const;    // TRUE if type is a singleton
   virtual bool empty(void) const;        // TRUE if type is vacuous
-  const jint _lo, _hi;          // Lower bound, upper bound
-
-  static const TypeInt *make(jint lo);
+  // A value is in the set represented by this TypeInt if it satisfies all
+  // the below constraints, see contains(jint)
+  const jint _lo, _hi;       // Lower bound, upper bound in the signed domain
+  const juint _ulo, _uhi;    // Lower bound, upper bound in the unsigned domain
```

Comment:
Can't we do without explicit fields to record unsigned hi / lo ?
We just need to present a unsigned view of signed _lo and _hi which can be done using safe macros.

Reply:
Ignoring the unsigned issue for a moment, and going back to Dual, if we had the concept of Complement, we could represent NON_ZERO as the complement of 0 <= x <= 0, which would be x > 0 || x < 0.  In general, the complement of lo <= x <= hi would be x > hi || x < lo, in contrast to the dual, which I believe is defined as the non-intuitive hi <= x <= lo.  I think complement would allow us to represent more complicated expressions, such as !(x>=lo && x<=hi).

If both dual and complement can be used to map between join and meet, then of the two complement seems more attractive and intuitive.  But maybe there is another reason we need dual than I'm missing.